### PR TITLE
TEST: clang build and clang-tidy

### PR DESCRIPTION
**Pull request by Sergei-Lebedev**
## What
Additional CI check that build UCC with clang and runs clang-tidy

## Why ?
clang-tidy is open source linter performing static code analysis. Here what it was able to find with current codebase
https://gist.github.com/Sergei-Lebedev/96141700ddd9de008802602c32ba780f
Warnings looks relevant

## How ?
Github actions runs clang build and clang-tidy

### CODE REVIEW

**Comment by manjugv** [comment]
@Sergei-Lebedev It seems useful. Do you see a lot of false warnings in your usage ?

File: src/core/ucc_lib.c [hunk 1]
```diff
@@ -27,16 +27,6 @@ static ucc_config_field_t ucc_lib_config_table[] = {
 UCC_CONFIG_REGISTER_TABLE(ucc_lib_config_table, "UCC", NULL, ucc_lib_config_t,
                           &ucc_config_global_list)
 
-static inline ucc_status_t ucc_cl_component_is_loaded(ucc_cl_type_t cl_type)
```
Comment (manjugv) [inline src/core/ucc_lib.c:30]: Is this related changes as well?

Comment (Sergei-Lebedev) [reply src/core/ucc_lib.c:30]: yes, clang says that function is unused and prints error during compilation

**Comment by Sergei-Lebedev** [comment]
@manjugv I need to look more carefully, but I think are no false warnings

**Comment by manjugv** [comment]
@Sergei-Lebedev Thanks. How do you think we should fix the warnings in the existing code? Do you already have fixes or do you want each of us to take ownership of some parts of the code?

**Comment by Sergei-Lebedev** [comment]
Fixes should be simple, clang mostly complains about potential memory leaks and invalid user input. I intentionally didn't fix warnings to show sample clang-tidy output.
For example following warring means we don't print error string if error happened
```
utils/ucc_component.c:55:10: warning: Although the value stored to 'error' is used in the enclosing expression, the value is never actually read from 'error' [clang-analyzer-deadcode.DeadStores]
    if ((error = dlerror()) != NULL) {
```
The question of this PR is what checks we want from clang-tidy, currently it is
```
-checks=-*,clang-analyzer-*,-clang-analyzer-security.insecureAPI.DeprecatedOrUnsafeBufferHandling
```
all checks from clang-analyzer group except checking for unsafe IO functions such sprintf

**Review by vspetrov — APPROVED** [review]

**Comment by Sergei-Lebedev** [comment]
@manjugv added .clang-tidy config file

<details>
  <summary>Enabled checks</summary>

clang-analyzer-apiModeling.StdCLibraryFunctions

clang-analyzer-apiModeling.TrustNonnull

clang-analyzer-apiModeling.google.GTest

clang-analyzer-apiModeling.llvm.CastValue

clang-analyzer-apiModeling.llvm.ReturnValue

clang-analyzer-core.CallAndMessage

clang-analyzer-core.CallAndMessageModeling

clang-analyzer-core.DivideZero

clang-analyzer-core.DynamicTypePropagation

clang-analyzer-core.NonNullParamChecker

clang-analyzer-core.NonnilStringConstants

clang-analyzer-core.NullDereference

clang-analyzer-core.StackAddrEscapeBase

clang-analyzer-core.StackAddressEscape

clang-analyzer-core.UndefinedBinaryOperatorResult

clang-analyzer-core.VLASize

clang-analyzer-core.builtin.BuiltinFunctions

clang-analyzer-core.builtin.NoReturnFunctions

clang-analyzer-core.uninitialized.ArraySubscript

clang-analyzer-core.uninitialized.Assign

clang-analyzer-core.uninitialized.Branch

clang-analyzer-core.uninitialized.CapturedBlockVariable

clang-analyzer-core.uninitialized.UndefReturn

clang-analyzer-cplusplus.InnerPointer

clang-analyzer-cplusplus.Move

clang-analyzer-cplusplus.NewDelete

clang-analyzer-cplusplus.NewDeleteLeaks

clang-analyzer-cplusplus.PlacementNew

clang-analyzer-cplusplus.PureVirtualCall

clang-analyzer-cplusplus.SelfAssignment

clang-analyzer-cplusplus.SmartPtrModeling

clang-analyzer-cplusplus.VirtualCallModeling

clang-analyzer-deadcode.DeadStores

clang-analyzer-fuchsia.HandleChecker

clang-analyzer-nullability.NullPassedToNonnull

clang-analyzer-nullability.NullReturnedFromNonnull

clang-analyzer-nullability.NullabilityBase

clang-analyzer-nullability.NullableDereferenced

clang-analyzer-nullability.NullablePassedToNonnull

clang-analyzer-nullability.NullableReturnedFromNonnull

clang-analyzer-optin.cplusplus.UninitializedObject

clang-analyzer-optin.cplusplus.VirtualCall

clang-analyzer-optin.mpi.MPI-Checker

clang-analyzer-optin.performance.GCDAntipattern

clang-analyzer-optin.performance.Padding

clang-analyzer-optin.portability.UnixAPI

clang-analyzer-security.FloatLoopCounter

clang-analyzer-security.insecureAPI.SecuritySyntaxChecker

clang-analyzer-security.insecureAPI.UncheckedReturn

clang-analyzer-security.insecureAPI.bcmp

clang-analyzer-security.insecureAPI.bcopy

clang-analyzer-security.insecureAPI.bzero

clang-analyzer-security.insecureAPI.decodeValueOfObjCType

clang-analyzer-security.insecureAPI.getpw

clang-analyzer-security.insecureAPI.gets

clang-analyzer-security.insecureAPI.mkstemp

clang-analyzer-security.insecureAPI.mktemp

clang-analyzer-security.insecureAPI.rand

clang-analyzer-security.insecureAPI.strcpy

clang-analyzer-security.insecureAPI.vfork

clang-analyzer-unix.API

clang-analyzer-unix.DynamicMemoryModeling

clang-analyzer-unix.Malloc

clang-analyzer-unix.MallocSizeof

clang-analyzer-unix.MismatchedDeallocator

clang-analyzer-unix.Vfork

clang-analyzer-unix.cstring.BadSizeArg

clang-analyzer-unix.cstring.CStringModeling

clang-analyzer-unix.cstring.NullArg

clang-analyzer-valist.CopyToSelf

clang-analyzer-valist.Uninitialized

clang-analyzer-valist.Unterminated

clang-analyzer-valist.ValistBase

clang-analyzer-webkit.NoUncountedMemberChecker

clang-analyzer-webkit.RefCntblBaseVirtualDtor

misc-redundant-expression

misc-static-assert

misc-unused-parameters

readability-redundant-control-flow
</details>

**Review by manjugv — APPROVED** [review]
